### PR TITLE
README.md: tutorials updated

### LINKED ISSUE #4: Improve README

### PR DESCRIPTION
**Pull request by mpecha**
README.md updated that corresponds to a current version of PermonSVM

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,17 +3,16 @@ PermonSVM - the PERMON SVM classifier
...
-GitHub: <https://github.com/permon/permonsvm>
 
 Please use [GitHub](https://github.com/permon/permonsvm) for issues and pull requests.
 
-Features
---------
+Feature overview
+-----------------
 
 - The scalable parallel solution for the linear C-SVM
```
Comment (jkruzik) [inline README.md:12]:
```suggestion
- The scalable parallel solution for the linear C-SVM 
+ Scalable (parallel) solution for the linear C-SVM
```

Comment (jkruzik) [inline README.md]:
```suggestion
Scalable (parallel) solution for the linear C-SVM 
```

File: README.md [hunk 2]
```diff
@@ -36,41 +35,90 @@ Features
...
-	- HDF5 (AIJ matrices support)
+	- HDF5 (AIJ and dense matrices)
 	- SVMLight
 
-Quick guide to PermonSVM installation
+Quick guide to installation
 -------------------------------------
 
-1. install [PermonQP](https://github.com/permon/permon)
+1. install [PermonQP](https://github.com/permon/permon) (follow instructions in the associated README.md)
```
Comment (jkruzik) [inline README.md]:
```suggestion
1. install [PermonQP](https://github.com/permon/permon) (follow instructions in the associated README)
```

File: README.md [hunk 3]
```diff
@@ -36,41 +35,90 @@ Features
...
 3. build PermonSVM simply using makefile (makes use of PETSc buildsystem):
    `make`
 4. if the build is successful, there is a new subdirectory named `$PETSC_ARCH` with the program library `$PETSC_ARCH/lib/libpermonsvm.{dylib,so,a}` and the executable `$PETSC_ARCH/bin/permonsvmfile`
    - shared library (.so) is built just if PETSc has been configured with option `--with-shared-libraries`
-   - all compiler settings are inherited from PETSc
+   - all compiler settings are inherited from PETSc and PermonQP
 
-Example of PermonSVM usage
+Tutorials
 --------------------------
```
Comment (jkruzik) [inline README.md:56]:
```suggestion
Tutorials illustrating basic functionality of the package are located in [`src/tutorials`](https://github.com/permon/permonsvm/tree/master/src/tutorials).

```

File: README.md [hunk 4]
```diff
@@ -36,41 +35,90 @@ Features
...
+
+	```bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_hyperopt 1 -svm_gs_logC_base 10 -svm_gs_logC_stride 1,2,1 -svm_nfolds 3 -cross_svm_warm_start 1 
+	-svm_cv_type stratified_kfold
+	```
+   
+### Using precomputed Gramian matrix
+
+PermonSVM uses an implicit representation of the Gramian matrix by default. Sometimes, it is reasonable to compute inner products related to the Gramian explicitly, typically, when a number of features is disproportionately larger than a number of samples. For such cases, PermonSVM provides functionality allowing to load precomputed Gramian matrix
```
Comment (jkruzik) [inline README.md]:
```suggestion
PermonSVM uses an implicit representation of the Gramian matrix by default. Sometimes, it is reasonable to compute inner products related to the Gramian explicitly, typically, when a number of features is disproportionately larger than a number of samples. For such cases, PermonSVM provides functionality allowing to load a precomputed Gramian matrix 
```

File: README.md [hunk 5]
```diff
@@ -36,41 +35,90 @@ Features
...
+### Using precomputed Gramian matrix
+
+PermonSVM uses an implicit representation of the Gramian matrix by default. Sometimes, it is reasonable to compute inner products related to the Gramian explicitly, typically, when a number of features is disproportionately larger than a number of samples. For such cases, PermonSVM provides functionality allowing to load precomputed Gramian matrix 
+
+```bash
+./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+-f_kernel $DATA_DIR/heart_scale.kernel.bin
+```
+
+The training dataset `src/tutorials/data/heart_scale` and testing dataset `src/tutorials/data/heart_scale.t` have been obtained by splitting the `heart_scale` dataset from the [LIBSVM dataset page](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html#heart). Other tutorials illustrating basic functionality of the package are located in `src/tutorials`.
```
Comment (jkruzik) [inline README.md]:
```suggestion
The training dataset `src/tutorials/data/heart_scale` and testing dataset `src/tutorials/data/heart_scale.t` have been obtained by splitting the `heart_scale` dataset from the [LIBSVM dataset page](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html#heart).
```

File: README.md (hunk 1 above)
Comment (mpecha) [reply README.md]: I am not sure why you suggest parallel in brackets.

File: README.md (hunk 3 above)
Comment (mpecha) [reply README.md:56]: Approved

File: README.md (hunk 5 above)
Comment (mpecha) [reply README.md]: Approved

File: README.md (hunk 4 above)
Comment (mpecha) [reply README.md]: Approved

File: README.md (hunk 2 above)
Comment (mpecha) [reply README.md]: Approved

File: README.md (hunk 1 above)
Comment (jkruzik) [reply README.md]: "The" is wrong. The rest does not matter. The parallel in brackets clarifies what type of scalability (parallel) you mean. For me the original meaning is - it is scalable (no idea in what sense) and parallel - and there is a missing comma between the adjectives.

See also 
> PETSc ...  routines for the scalable (parallel) solution of scientific applications (https://www.mcs.anl.gov/petsc/)

Comment (mpecha) [reply README.md]: I agree, article >> the << is wrong. Regarding the second part, thanks for the explanation.

**Review by jkruzik — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -36,42 +35,92 @@ Features
...
 	- Gini coefficient
-- Parallel loaders:
+- Parallel data loaders:
 	- PETSc binary
-	- HDF5 (AIJ matrices support)
+	- HDF5 (AIJ and dense matrices)
 	- SVMLight
 
-Quick guide to PermonSVM installation
+Quick guide to installation
```
Comment (haplav) [inline README.md]:
```suggestion
Quick installation guide
```

File: README.md [hunk 7]
```diff
@@ -36,42 +35,92 @@ Features
...
 -------------------------------------
 
-1. install [PermonQP](https://github.com/permon/permon)
+1. install [PermonQP](https://github.com/permon/permon) (follow instructions in the associated README)
```
Comment (haplav) [inline README.md]:
```suggestion
1. install [PermonQP](https://github.com/permon/permon) (follow instructions in its own [README.md](https://github.com/permon/permon/blob/master/README.md))
```

File: README.md [hunk 8]
```diff
@@ -36,42 +35,92 @@ Features
...
-   `./runsvmmpi 2 -f_training src/tutorials/data/heart_scale.bin -f_test src/tutorials/data/heart_scale.t.bin -svm_C 100`
-6. running PermonSVM on 2 MPI processes with C = 0.01 and _l2_ hinge loss function
-   `./runsvmmpi 2 -f_training src/tutorials/data/heart_scale.bin -f_test src/tutorials/data/heart_scale.t.bin -svm_loss_type L2 -svm_C 1e-2`
-7. running PermonSVM on 2 MPI processes with C = 0.01 and _l2_ hinge loss function (bias (standard) classification)
-	`./runsvmmpi 2 -f_training src/tutorials/data/heart_scale.bin -f_test src/tutorials/data/heart_scale.t.bin -svm_loss_type L2 -svm_C 1e-2 -svm_binary_mod 1`
+Tutorials illustrating basic functionality of the package are located in [`src/tutorials`](https://github.com/permon/permonsvm/tree/master/src/tutorials). We also provide the bash script [runsvmmpi](https://github.com/permon/permonsvm/tree/master/runsvmmpi) in the root directory of PermonSVM to easily run minimal working example [`src/bin/permonsvmfile.c`](https://github.com/permon/permonsvm/tree/master/src/bin/permonsvmfile.c). Several training and test datasets are located in ``DATA_DIR=src/tutorials/data``. Please set the variable before running following examples.
```
Comment (haplav) [inline README.md]:
```suggestion
* Tutorials illustrating basic functionality of the package are located in [`src/tutorials`](https://github.com/permon/permonsvm/tree/master/src/tutorials).
* We also provide the bash script [runsvmmpi](https://github.com/permon/permonsvm/tree/master/runsvmmpi) in the root directory of PermonSVM to easily run minimal working example [`src/bin/permonsvmfile.c`](https://github.com/permon/permonsvm/tree/master/src/bin/permonsvmfile.c).
* Several training and test datasets are located in `DATA_DIR=src/tutorials/data`.
* Please set the `DATA_DIR` variable before running following examples.
```

File: README.md [hunk 9]
```diff
@@ -36,42 +35,92 @@ Features
...
+
+### Using different classification methods
+
+1. running PermonSVM on 2 MPI processes with default settings (relaxed-bias classification, _l1_ hinge loss, C = 1, B = 1)
+   
+ 	```	bash
```
Comment (haplav) [inline README.md]:
```suggestion
 	```bash 
```

File: README.md [hunk 10]
```diff
@@ -36,42 +35,92 @@ Features
...
+	-svm_loss_type L2 -svm_C 1e-2 -svm_binary_mod 1
+	```
+	
+### Hyperparameter optimization
+
+1. running PermonSVM on 2 MPI processes with hyperparameter optimization with default settings (_l1_ hinge loss function, relaxed-bias classification, grid-search log2C = [-2:1:2], k-fold cross validation on 5 folds)
+
+	```	bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_hyperopt 1
```
Comment (haplav) [inline README.md]:
```suggestion
	```bash
	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
	  -svm_hyperopt 1
```

File: README.md [hunk 11]
```diff
@@ -36,42 +35,92 @@ Features
...
+	```
+   
+2. running PermonSVM on 2 MPI processes with grid-search on C = {0.1, 1, 10, 100} combined with cross validation on 3 folds that reuses a previous solution (warm start)
+
+	```bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_hyperopt 1 -svm_gs_logC_base 10 -svm_gs_logC_stride 1,2,1 -svm_nfolds 3 -cross_svm_warm_start 1
```
Comment (haplav) [inline README.md]:
```suggestion
	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
	  -svm_hyperopt 1 -svm_gs_logC_base 10 -svm_gs_logC_stride 1,2,1 -svm_nfolds 3 -cross_svm_warm_start 1
```

File: README.md [hunk 12]
```diff
@@ -36,42 +35,92 @@ Features
...
+	```
+  
+3. running PermonSVM on 2 MPI processes with grid search on C = {0.1, 1, 10, 100} and stratified k-fold cross validation on 3 folds with warm start
+
+	```bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_hyperopt 1 -svm_gs_logC_base 10 -svm_gs_logC_stride 1,2,1 -svm_nfolds 3 -cross_svm_warm_start 1 
+	-svm_cv_type stratified_kfold
```
Comment (haplav) [inline README.md]:
```suggestion
	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
	  -svm_hyperopt 1 -svm_gs_logC_base 10 -svm_gs_logC_stride 1,2,1 -svm_nfolds 3 -cross_svm_warm_start 1 \
	  -svm_cv_type stratified_kfold
```

File: README.md [hunk 13]
```diff
@@ -36,42 +35,92 @@ Features
...
+	```	bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_loss_type L2 -svm_C 1e-2
+	```
+  
+4. running PermonSVM on 2 MPI processes solving standard classification problem (binary mod 1), missclassification error quantification by _l2_ hinge loss, and C = 0.01
+	
+	```	bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin
+	-svm_loss_type L2 -svm_C 1e-2 -svm_binary_mod 1
```
Comment (haplav) [inline README.md]:
```suggestion
	```bash
	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
	  -svm_loss_type L2 -svm_C 1e-2 -svm_binary_mod 1
```

File: README.md [hunk 14]
```diff
@@ -36,42 +35,92 @@ Features
...
+	```	bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_C 100
+	```
+   
+3. running PermonSVM on 2 MPI processes with C = 0.01 and _l2_ hinge loss
+
+	```	bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_loss_type L2 -svm_C 1e-2
```
Comment (haplav) [inline README.md]:
```suggestion
	```bash
	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
	  -svm_loss_type L2 -svm_C 1e-2
```

File: README.md [hunk 15]
```diff
@@ -36,42 +35,92 @@ Features
...
+   
+ 	```	bash 
+ 	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin
+ 	```
+  
+2. running PermonSVM on 2 MPI processes with penalty parameter C = 100 
+	
+	```	bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_C 100
```
Comment (haplav) [inline README.md]:
```suggestion
	```bash
	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
	  -svm_C 100
```

File: README.md [hunk 16]
```diff
@@ -36,42 +35,92 @@ Features
...
+	-svm_cv_type stratified_kfold
+	```
+   
+### Using precomputed Gramian matrix
+
+PermonSVM uses an implicit representation of the Gramian matrix by default. Sometimes, it is reasonable to compute inner products related to the Gramian explicitly, typically, when a number of features is disproportionately larger than a number of samples. For such cases, PermonSVM provides functionality allowing to load precomputed Gramian matrix 
+
+```bash
+./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+-f_kernel $DATA_DIR/heart_scale.kernel.bin
```
Comment (haplav) [inline README.md]:
```suggestion
./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin \
  -f_kernel $DATA_DIR/heart_scale.kernel.bin
```

File: README.md [hunk 17]
```diff
@@ -36,42 +35,92 @@ Features
...
+
+	```bash
+	./runsvmmpi 2 -f_training $DATA_DIR/heart_scale.bin -f_test $DATA_DIR/heart_scale.t.bin 
+	-svm_hyperopt 1 -svm_gs_logC_base 10 -svm_gs_logC_stride 1,2,1 -svm_nfolds 3 -cross_svm_warm_start 1 
+	-svm_cv_type stratified_kfold
+	```
+   
+### Using precomputed Gramian matrix
+
+PermonSVM uses an implicit representation of the Gramian matrix by default. Sometimes, it is reasonable to compute inner products related to the Gramian explicitly, typically, when a number of features is disproportionately larger than a number of samples. For such cases, PermonSVM provides functionality allowing to load precomputed Gramian matrix
```
Comment (haplav) [inline README.md]:
```suggestion
PermonSVM uses an implicit representation of the Gramian matrix by default.
Sometimes, it is reasonable to compute inner products related to the Gramian explicitly, typically, when a number of features is disproportionately larger than a number of samples.
For such cases, PermonSVM provides functionality allowing to load precomputed Gramian matrix.
```
(Just newlines, I know it doesn't matter for the output, but the raw MD is more readable and it's better for versioning.)

**Review by haplav — APPROVED** [review]
@mpecha Looks good, just have a few typographic points.

**Comment by haplav** [comment]
@mpecha I suggest squash & merge once the suggestions are applied.
